File: app/types/expression-node.ts
import type { WhereClause } from './query-model'

/**
 * 式ノード（再帰的構造）
 *
 * SELECT句で使用できる式を表現する型
 */
export type ExpressionNode =
  | ColumnReference
  | LiteralValue
  | FunctionCall
  | SubqueryExpression
  | BinaryOperation
  | UnaryOperation

/**
 * カラム参照
 * 例: u.name, orders.total
 */
export interface ColumnReference {
  type: 'column'
  table?: string
  column: string
}

/**
 * リテラル値
 * 例: 'Hello', 123, true, null
 */
export interface LiteralValue {
  type: 'literal'
  valueType: 'string' | 'number' | 'boolean' | 'null'
  value: string | number | boolean | null
}

/**
 * 関数呼び出し
 * 例: UPPER(name), CONCAT(first_name, ' ', last_name)
 */
export interface FunctionCall {
  type: 'function'
  name: string
  category: 'string' | 'date' | 'numeric' | 'conditional' | 'aggregate'
  arguments: ExpressionNode[]
}

/**
 * サブクエリ式
 * 例: (SELECT COUNT(*) FROM orders WHERE user_id = u.id)
 */
export interface SubqueryExpression {
  type: 'subquery'
  query: SubqueryModel
}

/**
 * サブクエリモデル（簡易版）
 */
export interface SubqueryModel {
  select: ExpressionNode
  from: string
  where?: WhereClause
  alias?: string
}

/**
 * 二項演算
 * 例: price * quantity, age > 18
 */
export interface BinaryOperation {
  type: 'binary'
  operator:
    | '+'
    | '-'
    | '*'
    | '/'
    | '='
    | '!='
    | '>'
    | '<'
    | '>='
    | '<='
    | 'AND'
    | 'OR'
  left: ExpressionNode
  right: ExpressionNode
}

/**
 * 単項演算
 * 例: NOT active, -price
 */
export interface UnaryOperation {
  type: 'unary'
  operator: 'NOT' | '-'
  operand: ExpressionNode
}

File: src-tauri/src/models/expression_node.rs
use serde::{Deserialize, Serialize};

use crate::models::query::WhereClause;

/// 式ノード（再帰的構造）
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ExpressionNode {
    Column(ColumnReference),
    Literal(LiteralValue),
    Function(FunctionCall),
    Subquery(SubqueryExpression),
    Binary(BinaryOperation),
    Unary(UnaryOperation),
}

/// カラム参照
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ColumnReference {
    pub table: Option<String>,
    pub column: String,
}

/// リテラル値（型安全なenum版）
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "valueType", content = "value", rename_all = "lowercase")]
pub enum LiteralValue {
    String(String),
    Number(serde_json::Number),
    Boolean(bool),
    Null,
}

/// 関数呼び出し
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FunctionCall {
    pub name: String,
    pub category: FunctionCategory,
    pub arguments: Vec<ExpressionNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum FunctionCategory {
    String,
    Date,
    Numeric,
    Conditional,
    Aggregate,
}

/// サブクエリ式
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SubqueryExpression {
    pub query: Box<SubqueryModel>,
}

/// サブクエリモデル
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SubqueryModel {
    pub select: Box<ExpressionNode>,
    pub from: String,
    #[serde(rename = "where")]
    pub where_clause: Option<WhereClause>,
    pub alias: Option<String>,
}

/// 二項演算
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct BinaryOperation {
    pub operator: BinaryOperator,
    pub left: Box<ExpressionNode>,
    pub right: Box<ExpressionNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum BinaryOperator {
    #[serde(rename = "+")]
    Add,
    #[serde(rename = "-")]
    Subtract,
    #[serde(rename = "*")]
    Multiply,
    #[serde(rename = "/")]
    Divide,
    #[serde(rename = "=")]
    Equal,
    #[serde(rename = "!=")]
    NotEqual,
    #[serde(rename = ">")]
    GreaterThan,
    #[serde(rename = "<")]
    LessThan,
    #[serde(rename = ">=")]
    GreaterThanOrEqual,
    #[serde(rename = "<=")]
    LessThanOrEqual,
    #[serde(rename = "AND")]
    And,
    #[serde(rename = "OR")]
    Or,
}

impl BinaryOperator {
    pub fn as_str(&self) -> &'static str {
        match self {
            BinaryOperator::Add => "+",
            BinaryOperator::Subtract => "-",
            BinaryOperator::Multiply => "*",
            BinaryOperator::Divide => "/",
            BinaryOperator::Equal => "=",
            BinaryOperator::NotEqual => "!=",
            BinaryOperator::GreaterThan => ">",
            BinaryOperator::LessThan => "<",
            BinaryOperator::GreaterThanOrEqual => ">=",
            BinaryOperator::LessThanOrEqual => "<=",
            BinaryOperator::And => "AND",
            BinaryOperator::Or => "OR",
        }
    }
}

/// 単項演算
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct UnaryOperation {
    pub operator: UnaryOperator,
    pub operand: Box<ExpressionNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum UnaryOperator {
    #[serde(rename = "NOT")]
    Not,
    #[serde(rename = "-")]
    Negate,
}

impl UnaryOperator {
    pub fn as_str(&self) -> &'static str {
        match self {
            UnaryOperator::Not => "NOT",
            UnaryOperator::Negate => "-",
        }
    }
}
